pennylane/devices/experimental/device_api.py

```diff
@@ -0,0 +1,555 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the Abstract Base Class for the next generation of devices.
+"""
+# pylint: disable=comparison-with-callable
+import abc
+
+from numbers import Number
+from typing import Callable, Union, Sequence, Tuple, Optional
+
+from pennylane.tape import QuantumTape
+from pennylane import Tracker
+
+from .execution_config import ExecutionConfig, DefaultExecutionConfig
+
+QuantumTapeBatch = Sequence[QuantumTape]
+QuantumTape_or_Batch = Union[QuantumTape, QuantumTapeBatch]
+
+
+# pylint: disable=unused-argument, no-self-use
+class Device(abc.ABC):
+    """A device driver that can control one or more backends. A backend can be either a physical
+    Quantum Processing Unit or a virtual one such as a simulator.
+
+    .. warning::
+
+        This interface is **experimental** and not yet integrated with the rest of PennyLane.
+
+    Device drivers should be configured to run under :func:`~.enable_return`, the newer
+    return shape specification.
+
+    Only the ``execute`` method must be defined to construct a device driver.
+
+    .. details::
+        :title: Design Motivation
+
+        **Streamlined interface:** Only methods that are required to interact with the rest of PennyLane will be placed in the
+        interface. Developers will be able to clearly see what they can change while still having a fully functional device.
+
+        **Reduction of duplicate methods:** Methods that solve similar problems are combined together. Only one place will have
+        to solve each individual problem.
+
+        **Support for dynamic execution configurations:** Properties such as shots belong to specific executions.
+
+        **Greater coverage for differentiation methods:** Devices can define any order of derivative, the vector jacobian product,
+        or the jacobian vector product.  Calculation of derivatives can be done at the same time as execution to allow reuse of intermediate
+        results.
+
+    .. details::
+        :title: Porting from the old interface
+
+        :meth:`~.Device.batch_execute` and :meth:`~.Device.execute` are now a single method, :meth:`~.Device.execute`
+
+        :meth:`~.Device.batch_transform` and :meth:`~.Device.expand_fn` are now a single method, :meth:`~.Device.preprocess`
+
+        Shot information is no longer stored on the device, but instead specified on individual input :class:`~.QuantumTape`.
```

Comment:
I suppose we can change this later if needed, but my understanding was that shot information would be _optionally set_ on a tape. It could be omitted from the tape, then the executing device would get it from the executionconfig. otherwise we could just default to writing myEC.shots directly onto each tape and then not have to decide where to get shots from down the road.

Reply:
We can easily resolve the location of shots to the most specific location in "Core" code before its sent to the device. Then the device won't need to resolve priority.  That will be the responsibility of a preprocessing step.  But this will be more fleshed out over the next few months.